s3/src/main/java/io/ddf/s3/S3DDF.java

```diff
@@ -0,0 +1,135 @@
+package io.ddf.s3;
+
+import io.ddf.DDF;
+import io.ddf.DDFManager;
+import io.ddf.content.Schema;
+import io.ddf.content.SqlResult;
+import io.ddf.datasource.DataFormat;
+import io.ddf.exception.DDFException;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by jing on 12/2/15.
+ */
+public class S3DDF extends DDF {
+    // Whether this ddf has header.
+    private Boolean mHasHeader = false;
+    // It's a directory or file.
+    private Boolean mIsDir;
+    // The format of this s3ddf. If it's a folder, we requires that all the files in the folder should have the same
+    // format, otherwise the dataformat will be set to the dataformat of the first file under this folder.
+    private DataFormat mDataFormat;
+    // Schema String.
+    private String mSchemaString;
+    // Bucket.
+    private String mBucket;
+    // Path after bucket.
+    private String mKey;
+
+    /**
+     * S3DDF is the ddf for s3. It point to a single S3DDFManager, and every S3DDF is a unqiue mapping to a s3 uri.
+     * The schema should store the s3 uri as tablename.
+     */
+
+    public S3DDF(S3DDFManager manager, String path, String schema) throws DDFException {
+        super(manager, null, null, null, null, null);
+        mSchemaString = schema;
+        this.getBucketAndPath(path);
+        checkStatus();
+    }
+
+    public S3DDF(S3DDFManager manager, String path) throws DDFException {
+        super(manager, null, null, null, null, null);
+        this.getBucketAndPath(path);
+        checkStatus();
+    }
+
+    public S3DDF(S3DDFManager manager, String bucket, String pathNoBucket, String schema) throws DDFException {
+        super(manager, null, null, null, null, null);
+        mBucket = bucket;
+        mKey = pathNoBucket;
+        mSchemaString = schema;
+        checkStatus();
+    }
+
+
+    /**
+     * @brief Get the bucket and path out of a given uri.
+     * @param path
+     * @return
+     */
+    private void getBucketAndPath(String path) {
+        int firstSlash = path.indexOf('/');
+        mBucket = path.substring(0, firstSlash);
+        mKey = path.substring(firstSlash + 1);
+    }
+
+    public void checkStatus() throws DDFException {
+        // Check directory or file.
+        S3DDFManager s3DDFManager = (S3DDFManager)this.getManager();
```

Comment:
Maybe override getManager() in this class to return S3DDFManager. Java permits this degree of interface specialization.